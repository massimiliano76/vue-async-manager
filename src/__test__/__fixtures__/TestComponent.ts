import { CreateElement } from 'vue'

export default {
  name: 'TestComponent',
  render(h: CreateElement) {
    return h('div', { class: { 'test-component': true } }, 'test component')
  }
}
