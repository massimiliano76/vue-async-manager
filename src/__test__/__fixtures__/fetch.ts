import { Component } from 'vue'

export default (mockData?: any, throwError?: any) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      !throwError
        ? resolve(
            mockData || {
              code: 0,
              msg: 'success',
              data: {}
            }
          )
        : reject('request error')
    }, 1000)
  })
}

export function dynamicImport(
  Comp: Component,
  throwError: boolean = false
): Promise<Component> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      throwError ? reject('load error') : resolve(Comp)
    }, 1000)
  })
}
