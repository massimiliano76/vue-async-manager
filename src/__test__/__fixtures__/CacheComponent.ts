import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

export default {
  name: 'CacheComponent',
  created() {
    this.$rm = createResource(() => fetch())
    // This.promiser is used in test cases, waiting for resource requests to complete
    this.promiser = this.$rm.read()
  },
  render(h: CreateElement) {
    return h(
      'div',
      { class: { 'cache-component': true } },
      JSON.stringify(this.$rm.$result)
    )
  }
}
