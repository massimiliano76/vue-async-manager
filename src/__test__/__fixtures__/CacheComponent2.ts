import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

export default {
  name: 'CacheComponent2',
  data() {
    return {
      res: {}
    }
  },
  async created() {
    this.resource = createResource(() => fetch())
    // This.promiser is used in test cases, waiting for resource requests to complete
    this.promiser = this.resource.read()
    await this.promiser
    this.res = this.resource.$res.$$result
  },
  render(h: CreateElement) {
    return h(
      'div',
      { class: { 'cache-component-2': true } },
      JSON.stringify(this.res)
    )
  }
}
