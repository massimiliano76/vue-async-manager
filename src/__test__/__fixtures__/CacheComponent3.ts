import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

const resource = createResource(() => fetch())

export default {
  name: 'CacheComponent3',
  data() {
    return {
      res: {}
    }
  },
  async created() {
    const res = resource.read()
    // This.promiser is used in test cases, waiting for resource requests to complete
    this.promiser = res.$$waiter
    await res.$$waiter
    this.res = res.$$result
  },
  render(h: CreateElement) {
    console.log('CacheComponent render')
    return h(
      'div',
      { class: { 'cache-component-3': true } },
      JSON.stringify(this.res)
    )
  }
}
