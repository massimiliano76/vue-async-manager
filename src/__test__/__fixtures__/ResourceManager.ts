import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

export default {
  name: 'ResourceManager',
  created() {
    this.$rm = createResource((mockData: any) => fetch(mockData))
    this.$rm2 = this.$rm.fork()

    this.promiser1 = this.$rm.read({ name: 'foo' })
    this.promiser2 = this.$rm2.read({ name: 'bar' })
  },
  render(h: CreateElement) {
    return h('div')
  }
}
