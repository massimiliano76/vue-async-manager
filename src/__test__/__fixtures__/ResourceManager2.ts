import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

export default {
  name: 'ResourceManager',
  created() {
    this.$rm = createResource((mockData: any) => fetch(mockData), {
      prevent: true
    })

    this.promiser1 = this.$rm.read({ name: 'foo' })
    this.promiser2 = this.$rm.read({ name: 'bar' })

    this.$rm2 = createResource((mockData: any) => fetch(mockData))

    this.promiser3 = this.$rm2.read({ name: 'foo' })
    this.promiser4 = this.$rm2.read({ name: 'bar' })
  },
  render(h: CreateElement) {
    return h('div')
  }
}
