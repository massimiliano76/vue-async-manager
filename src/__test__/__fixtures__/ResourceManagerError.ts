import { CreateElement } from 'vue'
import createResource from '../../createResource'
import fetch from './fetch'

export default {
  name: 'ResourceManager',
  created() {
    this.$rm = createResource((mockData: any) =>
      fetch(mockData, true /* throw error */)
    )

    this.promiser = this.$rm.read({ name: 'foo' })
  },
  render(h: CreateElement) {
    return h('div')
  }
}
