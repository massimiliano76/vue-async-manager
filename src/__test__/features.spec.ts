import Vue from 'vue'
import installer from '../index'
import lazy from '../lazy'
import TestComp from './__fixtures__/TestComponent'
import { dynamicImport } from './__fixtures__/fetch'
import { SSVue } from '../Suspense'

Vue.config.devtools = false
Vue.config.productionTip = false

Vue.use(installer)

describe('Lazy Error: ', () => {
  test('Lazy component failed to load', async () => {
    const TestComponent = lazy(() =>
      dynamicImport(TestComp, true /* throw Error */)
    )

    const ins = new Vue({
      components: {
        TestComponent
      },
      render(h) {
        return h('Suspense', [
          h('div', { class: { fallback: true }, slot: 'fallback' }, 'loading'),
          h('div', { class: { error: true }, slot: 'error' }, 'error'),
          h('TestComponent'),
          h('p', 'static')
        ])
      }
    })

    ins.$mount()

    expect(ins.$el.outerHTML).toMatchSnapshot()

    try {
      await (ins.$children[0] as SSVue).promiser
    } catch (err) {
      expect(ins.$el.outerHTML).toMatchSnapshot()
    }
  })
})

describe('Suspense event: ', () => {
  test('The resolved event should be triggered correctly', async () => {
    const TestComponent = lazy(() => dynamicImport(TestComp))

    const onResolved = jest.fn()
    const onRejected = jest.fn()

    const ins = new Vue({
      components: {
        TestComponent
      },
      render(h) {
        return h(
          'Suspense',
          {
            on: { resolved: onResolved, rejected: onRejected }
          },
          [h('TestComponent')]
        )
      }
    })

    ins.$mount()

    await (ins.$children[0] as SSVue).promiser

    expect(onResolved.mock.calls.length).toBe(1)
    expect(onRejected.mock.calls.length).toBe(0)
  })

  test('The rejected event should be triggered correctly', async () => {
    const TestComponent = lazy(() =>
      dynamicImport(TestComp, true /* throw Error */)
    )

    const onResolved = jest.fn()
    const onRejected = jest.fn()

    const ins = new Vue({
      components: {
        TestComponent
      },
      render(h) {
        return h(
          'Suspense',
          {
            on: { resolved: onResolved, rejected: onRejected }
          },
          [h('TestComponent')]
        )
      }
    })

    ins.$mount()

    try {
      await (ins.$children[0] as SSVue).promiser
    } catch (err) {
      expect(onResolved.mock.calls.length).toBe(0)
      expect(onRejected.mock.calls.length).toBe(1)
    }
  })
})
