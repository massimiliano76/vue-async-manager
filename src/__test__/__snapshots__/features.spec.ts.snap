// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy Error:  Lazy component failed to load 1`] = `"<div class=\\"vue-suspense-wrapper\\"><!----><p>static</p><div class=\\"fallback\\">loading</div></div>"`;

exports[`Lazy Error:  Lazy component failed to load 2`] = `"<div class=\\"vue-suspense-wrapper\\"><div class=\\"error\\">error</div></div>"`;
